File: 0410-split-array-largest-sum/0410-split-array-largest-sum.java
class Solution {
        public int splitArray(int[] nums, int k) {
        int start = 0;
        int end = 0;

        for (int i = 0; i < nums.length; i++) {
            start = Math.max(start,nums[i]); // in the end of loop this will contain max element index
            end += nums[i];
        }

        // binary search
        while(start<end){
            // try for the middle as potential answer
            int mid = start + (end-start)/2;

            // calculate how many pieces you can divide this with max sum
            int sum = 0;
            int pieces = 1;
            for (int num:nums) {
                if (sum + num > mid) {
                    // you can't add this to subarray make new one
                    // say add this in new subarray, then num = sum
                    sum = num;
                    pieces++;
                } else {
                    sum += num;
                }
            }

            if(pieces>k){
                start = mid+1;
            }else{
                end = mid;
            }

        }
        return end; // here start is equal to end
    }
}

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
class Solution {
public:
    int max(vector<int>& arr){
    int maxi = arr[0];
    for (int i = 1; i < arr.size(); i++)
    {
        if(arr[i]>maxi) maxi = arr[i];
    }
    return maxi;
}

int sum(vector<int>& arr){
    int sum = 0;
    for (int i = 0; i < arr.size(); i++)
    {
        sum += arr[i];
    }
    return sum;
}

int calculateStudents(vector<int>& arr, int pages){
    int student = 1;
    int pagesPerStudent = 0;
    for (int i = 0; i < arr.size(); i++)
    {
        if(pagesPerStudent + arr[i] <= pages){
            pagesPerStudent += arr[i];
        }else{
            student++;
            pagesPerStudent = arr[i];
        }
    }
    return student;
}

int findPages(vector<int>& arr, int n, int m){
    if (m > arr.size()) return -1;
    int low = max(arr);
    int high = sum(arr);
    while(low<=high){
        int mid = (low+high)/2;
        if(calculateStudents(arr,mid)>m){
            low = mid+1;
        }else{
            high = mid-1;
        }
    }
    return low;
}
    int splitArray(vector<int>& nums, int k) {
        return findPages(nums, nums.size(), k);
    }
};